overlord/devicestate/devicemgr.go

```diff
@@ -954,6 +957,50 @@ func (m *DeviceManager) Ensure() error {
 	return nil
 }
 
+// withKeypairMgr invokes a function making the device KeypairManager
+// available to it.
+// It uses the right location for the manager depending on UC16/18 vs 20,
+// the latter uses ubuntu-save.
+// For UC20 it also checks that ubuntu-save is mounted.
+func (m *DeviceManager) withKeypairMgr(f func(asserts.KeypairManager) error) error {
+	// we use the model to check whether this is a UC20 device
+	// TODO: during a theoretical UC18->20 remodel the location of
+	// keypair manager keys would move, we will need dedicated code
+	// to deal with that, this code typically will return the old location
+	// until a restart
+	model, err := m.Model()
+	if err == state.ErrNoState {
+		return fmt.Errorf("internal error: cannot access device keypair manager before a model is set")
+	}
+	if err != nil {
+		return err
+	}
+	underSave := false
+	if model.Grade() != asserts.ModelGradeUnset {
+		// on UC20 the keys are kept under the save dir
+		underSave = true
+	}
+	where := dirs.SnapDeviceDir
+	if underSave {
+		// check for availability in case save gets mounted/unmounted,
+		// it's responsibility of the caller to have it mounted
+		if !m.saveAvailable {
```

Comment:
Does that mean that ubuntu-save is indeed mandatory for UC20 now?

Reply:
no, it means though that some things will always go under /save, whether it comes from the partition or its just the dir in daa. If it comes from the partition it needs to be mounted.  The idea is that when/if we mount/unmount ubuntu-save, saveAvailable goes true or false depending,  if it's just the dir it's always true.

Reply:
Sorry, got confused with by the last commit and didn't notice that we set it to true in the branch when there is no mount too.